helsingborg-stad/app-mitt-helsingborg
# Feat/encryption system rework with password

## Explain the changes you’ve made
* Added a new encryption system that works with abstract "strategies" for encryption and decryption.
  * Two strategies are included; DeviceLocal (same as old "private AES" used for solo applicants), and Password-based (will replaced Symmetric Key)
* Deprecated existing encryption service and helper

Note: new encryption system is not being used by anything in this PR. Old encryption service is still being used.

## Explain why these changes are made

The current encryption system is not extensible enough and making changes are hard. This PR change was needed to facilitate implementing Password-based encryption ("pin code") as well as preparing to remove Symmetric Key exchange encryption.

## Explain your solution

Broadly two things have happened:

* The now deprecated `EncryptionService` and `EncryptionHelper` has been moved to `source/services/encryption/DEPRECATED` and files using them has been updated to point there (`CaseOverview` etc.).
* New files has been added to `source/services/encryption` that implements the new system, that are not used by anything yet.

### The new encryption system
Each method of encryption implements `IEncryptionStrategy`. They are oblivious to project-specific data structures (`Case`, `Form` etc.) and work only with raw data and abstract interfaces.

`CaseEncryptionService` is the glue between the generic strategies and the project-specific data. It handles encryption/decryption based on `Case` objects and `Form` objects. This is generally the class that will be called on by the UI layer or contexts.

`CaseEncryptionHelper` includes a set of commonly used types and functions as well as manages the mapping between `Case` encryption type and concrete encryption strategy (might move the second part to a separate concern?).

## How to test

Concrete example:

1. Checkout this branch
2. Make sure tests pass
3. Make sure app works as usual (start new case, encrypt ongoing, submit etc.)

## Tested environments

- [] Storybook on a iOS device/simulator.
- [] Building the Application on a Android device/simulator.

## Screenshots

## Other notes


---

## Checklist for reviewers

Here are a few points that are useful for reviewers to check.

The short list of the (maybe) most important points are:

- Does this PR work as expected?
- Does the PR properly only concern a single feature/bug/thing?
- Is the code clear and concise?
- Will the changes impact existing functionality?
- Are there any linting errors (ESLint) that should be resolved?
- Do the automatic GitHub checks pass?

However it is good if a reviewer can be as thorough as possible,
within reasonable limits of course.

### Functionality

- Does this PR work as expected?
- If this is a new feature, does it add value, or is it a sign of feature-creep?
- Will the changes impact existing functionality?
- Does the changes create inconsistencies somewhere else?
- Does the code work on different platforms if applicable?
- How does the code handle if unexpected values are passed to functions etc.?

### Readability, syntax, formatting, style

- Is the code clear and concise?
- Has styling rules (Prettier) been applied to included files?
- Are functions/variables/etc. aptly named?

### Design

- Is the code properly planned and designed?
- Is code organized/separated in a logical fashion?
- Does the code follow clean code practices to a reasonable degree?
- Is the code self-documenting enough?
- Is the code in sync with existing patterns/technologies?
- Does the code reuse existing functionality where applicable?
- Is the code designed with reusability in mind?
- Does the code have hard-coded values that can be refactored?

### Patterns, idioms, tests, best practices

- Are there any linting errors (ESLint) that should be resolved?
- Is the code properly typed (TypeScript)?
- Does the code utilize language features to their full potential?

### Debugging, testing, configuring

- Is appropriate logging used?
- Have tests been added/modified to accommodate the changes?
- Is the feature sufficiently configurable in case tweaks are required in the future?

### Performance, reliability, scalability

- Is the code well optimized?
- Does the code scale?
- Does the code handle failure scenarios?

### Security

- Is there any obvious potential for exploits and security threats?
- Does code check proper authorization where applicable?

### Etiquettes

- Are commits atomic (does only one thing per commit)?
- Does the PR properly only concern a single feature/bug/thing?
- Are commit messages well-written?
- Do the automatic GitHub checks pass?
